File: .agents/skills/weaver-oa-dev/scripts/api_search.js
#!/usr/bin/env node
/**
 * 泛微OA (Ecology 9) 全量 538 个后端 REST API 快速检索工具 (Node.js 版)
 * 纯原生 Node.js 实现，零第三方依赖，绿色免安装
 *
 * 运行方式:
 *   node scripts/api_search.js 待办
 *   node scripts/api_search.js -m 工作流程
 *   node scripts/api_search.js -m 人力资源 -X POST
 */

const fs = require('fs');
const path = require('path');

// 统一采用相对当前脚本的本地路径，杜绝任何外部或绝对路径依赖
const RAW_JSON_PATH = path.join(__dirname, 'apis_dictionary.json');

function loadApis() {
  if (fs.existsSync(RAW_JSON_PATH)) {
    return JSON.parse(fs.readFileSync(RAW_JSON_PATH, 'utf8'));
  }
  console.error('错误: 找不到 API 索引文件 ' + RAW_JSON_PATH);
  return [];
}

function searchApis({ keyword = null, moduleName = null, method = null, limit = 20 }) {
  const apis = loadApis();
  const kw = keyword ? keyword.toLowerCase().trim() : null;
  const mod = moduleName ? moduleName.toLowerCase().trim() : null;
  const meth = method ? method.toUpperCase().trim() : null;

  const results = [];

  for (const api of apis) {
    const apiModule = api.module || '';
    const apiName = api.apiName || '';
    const apiPath = api.path || '';
    const apiMethod = api.method || '';

    if (mod && !apiModule.toLowerCase().includes(mod)) {
      continue;
    }

    if (meth && apiMethod.toUpperCase() !== meth) {
      continue;
    }

    if (kw) {
      const matchName = apiName.toLowerCase().includes(kw);
      const matchPath = apiPath.toLowerCase().includes(kw);
      const matchModule = apiModule.toLowerCase().includes(kw);
      if (!matchName && !matchPath && !matchModule) {
        continue;
      }
    }

    results.push(api);
  }

  console.log(`\n共检索到 ${results.length} 个匹配接口 (展示前 ${Math.min(limit, results.length)} 条):\n`);
  console.log(`${'序号'.padEnd(4)} | ${'模块'.padEnd(10)} | ${'方法'.padEnd(6)} | ${'接口路径'.padEnd(52)} | 接口名称`);
  console.log('-'.repeat(105));

  results.slice(0, limit).forEach((r, idx) => {
    const indexStr = String(idx + 1).padEnd(4);
    const modStr = r.module.padEnd(10);
    const methStr = r.method.padEnd(6);
    const pathStr = r.path.padEnd(52);
    console.log(`${indexStr} | ${modStr} | ${methStr} | ${pathStr} | ${r.apiName}`);
  });

  console.log('-'.repeat(105));
  if (results.length > limit) {
    console.log(`提示: 还有 ${results.length - limit} 条结果未展示，可通过缩小检索范围查看。\n`);
  } else {
    console.log('');
  }
}

function main() {
  const args = process.argv.slice(2);
  let keyword = null;
  let moduleName = null;
  let method = null;
  let limit = 20;

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '-m' || arg === '--module') {
      moduleName = args[++i];
    } else if (arg === '-X' || arg === '--method') {
      method = args[++i];
    } else if (arg === '-n' || arg === '--limit') {
      limit = parseInt(args[++i], 10) || 20;
    } else if (!arg.startsWith('-')) {
      keyword = arg;
    }
  }

  if (!keyword && !moduleName && !method) {
    console.log('泛微OA 后端 REST API 快速检索工具 (Node.js 版)');
    console.log('用法:');
    console.log('  node scripts/api_search.js <关键词>');
    console.log('  node scripts/api_search.js -m <模块名> [-X <HTTP方法>]');
    console.log('\n示例:');
    console.log('  node scripts/api_search.js 待办');
    console.log('  node scripts/api_search.js -m 工作流程');
    console.log('  node scripts/api_search.js -m 人力资源 -X POST -n 10');
    return;
  }

  searchApis({ keyword, moduleName, method, limit });
}

main();

File: .agents/skills/weaver-oa-dev/scripts/api_search.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
泛微OA (E-Cology 9) 538 个后端 REST API 快速检索工具
支持按关键词（中文名、URL路径、参数名、模块名）进行精确或模糊检索
"""

import os
import sys
import json
import io
import argparse

if sys.platform == 'win32':
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
SCRATCH_JSON = os.path.join(r"C:\Users\Admin\.gemini\antigravity\brain\86c241ed-45ea-453f-a153-6d579096f1df\scratch\all_apis_raw.json")

def load_apis():
    if os.path.exists(SCRATCH_JSON):
        with open(SCRATCH_JSON, 'r', encoding='utf-8') as f:
            return json.load(f)
    print("错误: 找不到 API 数据源文件")
    return []

def search_apis(keyword=None, module=None, method=None, limit=20):
    docs = load_apis()
    results = []
    
    kw = keyword.lower().strip() if keyword else None
    mod = module.lower().strip() if module else None
    meth = method.upper().strip() if method else None

    for item in docs:
        data = item['data']
        leaf = item['leaf']
        
        title = data.get('title') or leaf.get('name') or ''
        url = data.get('url', '')
        category = leaf.get('category', '')
        full_path = leaf.get('full_path', '')
        m = (data.get('method') or data.get('methodSelect') or ('POST' if 'paService' in url else 'GET')).upper()
        memo = data.get('memo', '')
        params = [p.get('name', '') for p in data.get('params', [])]

        if mod and mod not in category.lower() and mod not in full_path.lower():
            continue
        if meth and meth != m:
            continue
        if kw:
            match = (
                kw in title.lower() or
                kw in url.lower() or
                kw in memo.lower() or
                any(kw in p.lower() for p in params)
            )
            if not match:
                continue

        results.append({
            'title': title,
            'url': url,
            'method': m,
            'module': category,
            'path': full_path,
            'params_count': len(params),
            'memo': memo
        })

    print(f"\n共检索到 {len(results)} 条匹配接口 (展示前 {min(limit, len(results))} 条):\n")
    print(f"{'序号':<4} | {'模块':<8} | {'Method':<6} | {'接口名称':<30} | {'Endpoint 路径'}")
    print("-" * 100)
    for idx, r in enumerate(results[:limit], 1):
        print(f"{idx:<4} | {r['module']:<8} | {r['method']:<6} | {r['title'][:28]:<30} | {r['url']}")
    print("-" * 100)
    if len(results) > limit:
        print(f"提示: 还有 {len(results) - limit} 条结果未展示，可通过 --limit 参数增加展示数量。")

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description="泛微OA E9 API 快速检索工具")
    parser.add_argument("keyword", nargs="?", default=None, help="搜索关键词 (支持接口名、URL路径、参数名)")
    parser.add_argument("-m", "--module", help="指定模块筛选 (如: 流程, 人力, 考勤, 知识, 表单, 门户)")
    parser.add_argument("-X", "--method", help="指定 HTTP Method (GET, POST)")
    parser.add_argument("-n", "--limit", type=int, default=20, help="最多显示条数 (默认 20)")
    
    args = parser.parse_args()
    if not args.keyword and not args.module and not args.method:
        parser.print_help()
        print("\n示例: python api_search.py 待办")
        print("示例: python api_search.py -m 人力资源 -X POST")
        print("示例: python api_search.py /api/workflow")
    else:
        search_apis(args.keyword, args.module, args.method, args.limit)
